radix-common/src/data/manifest/model/manifest_resource_assertion.rs

```diff
@@ -200,6 +200,121 @@ impl ManifestResourceConstraint {
+                        actual_amount: amount,
+                    });
+                }
+            }
+            ManifestResourceConstraint::ExactNonFungibles(expected_exact_ids) => {
+                if !expected_exact_ids.eq(&ids) {
+                    return Err(ResourceConstraintError::ExactNonFungibles {
+                        expected_exact_ids: Box::new(expected_exact_ids),
+                        actual_ids: Box::new(ids),
+                    });
+                }
+            }
+            ManifestResourceConstraint::AtLeastNonFungibles(expected_at_least_ids) => {
+                if !expected_at_least_ids.is_subset(&ids) {
+                    return Err(ResourceConstraintError::AtLeastNonFungibles {
+                        actual_ids: Box::new(ids),
+                        expected_at_least_ids: Box::new(expected_at_least_ids.clone()),
+                    });
+                }
+            }
+            ManifestResourceConstraint::General(constraint) => {
+                constraint
+                    .validate_non_fungible(&ids)
+                    .map_err(ResourceConstraintError::General)?;
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn validate_fungible(self, amount: Decimal) -> Result<(), ResourceConstraintError> {
+        match self {
+            ManifestResourceConstraint::NonZeroAmount => {
+                if amount.is_zero() {
+                    return Err(ResourceConstraintError::NonZeroAmount);
+                }
+            }
+            ManifestResourceConstraint::ExactAmount(expected_exact_amount) => {
+                if amount.ne(&expected_exact_amount) {
+                    return Err(ResourceConstraintError::ExactAmount {
+                        actual_amount: amount,
+                        expected_exact_amount,
+                    });
+                }
+            }
+            ManifestResourceConstraint::AtLeastAmount(expected_at_least_amount) => {
+                if amount < expected_at_least_amount {
+                    return Err(ResourceConstraintError::AtLeastAmount {
+                        expected_at_least_amount,
+                        actual_amount: amount,
+                    });
+                }
+            }
+            ManifestResourceConstraint::ExactNonFungibles(..) => {
+                return Err(ResourceConstraintError::ExpectedNonFungibleResourceButIsFungible);
+            }
+            ManifestResourceConstraint::AtLeastNonFungibles(..) => {
+                return Err(ResourceConstraintError::ExpectedNonFungibleResourceButIsFungible);
+            }
+            ManifestResourceConstraint::General(constraint) => {
+                constraint
+                    .validate_fungible(amount)
+                    .map_err(ResourceConstraintError::General)?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, ScryptoSbor)]
+pub enum ResourceConstraintError {
+    NonZeroAmount,
```

Comment:
MINOR: Like [in this comment](https://github.com/radixdlt/radixdlt-scrypto/pull/1947/files/e0c1455f6296a658bd98992ff932e8f8ca648078#r1790136887), I'm still not sure the error names are particularly descriptive of error states.

Perhaps prepending `Expected` might be helpful? e.g. `ExpectedNonZeroAmount` / `ExpectedExactAmount`